Add script name to resource for Sinatra if available

delner opened the pull request:
This pull request prepends `request.script_name` to Sinatra trace resource names, if available and enabled. e.g. `get '/endpoint'` from a Sinatra app rooted in Rack via `map '/foo'` would change the trace `resource` from `/endpoint` to `/foo/endpoint`.

This new behavior is disabled by default. It can be enabled by adding the `resource_script_names` flag to the configuration. e.g.

```
Datadog.configure do |c|
  c.use :sinatra, resource_script_names: true
end
```

Or via `Datadog.configuration[:sinatra][:resource_script_names] = true`

This flag is useful for particular cases where there are multiple Sinatra apps configured in a `config.ru` that have overlapping endpoints rooted at different namespaces. Take for example the following app:

**config.ru:**

```
require 'bundler'
Bundler.require

require './lib/sinatra/app'

map '/foo' do
  run FooApp
end

map '/bar' do
  run BarApp
end
```

**lib/sinatra/app.rb:**

```
require 'ddtrace'
require 'ddtrace/contrib/sinatra/tracer'
require 'sinatra'

require_relative 'app/foo'
require_relative 'app/bar'
```

**lib/sinatra/app/foo.rb:**

```
class FooApp < Sinatra::Base
  register Datadog::Contrib::Sinatra::Tracer

  get '/endpoint' do
    'Foo!'
  end
end
```

**lib/sinatra/app/bar.rb:**

```
class BarApp < Sinatra::Base
  register Datadog::Contrib::Sinatra::Tracer

  get '/endpoint' do
    'Bar!'
  end
end
```

If you were to call `GET /foo/endpoint`, then `GET /bar/endpoint`, you would hit each of the apps correctly on their respective paths. However, the two traces created from these requests would report the same resource as `/endpoint` in the UI, which could lead to some confusion (since they are not the same app.)

By adding `request.script_name` onto the front of a path (see [Rack specification](http://www.rubydoc.info/github/rack/rack/file/SPEC) for more info), it allows these traces to be disambiguated as `GET /foo/endpoint` and `GET /bar/endpoint` for a less confusing user experience.

p-lambert commented on test/contrib/sinatra/multi_app_test.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  def setup
+    @writer = FauxWriter.new()
```
Can we remove the empty `()`? 🙏

delner replied:
Yes, we can!

p-lambert commented on lib/ddtrace/contrib/sinatra/tracer.rb:
```diff
@@ -34,7 +34,7 @@ def route(verb, action, *)
           # incoming request.
           condition do
-            @datadog_route = action
+            @datadog_route = "#{request.script_name}#{action}"
```
This feature seems kind of an advanced use-case IMO. Could we prepend this conditionally, using a configuration param? I think for most users the resource name without the script name would still be a better default.

delner replied:
I think that's reasonable. Any thoughts about what you'd call it?

p-lambert replied:
`script_name_prefix`?

p-lambert commented:
Very good PR description, by the way :)

delner commented:
Okay, modified per feedback. I added a `resource_script_names` flag to the Sinatra tracer, which defaults to `false`, disabling this new behavior by default. Also added tests to cover this flag.
